src/rpc/blockchain.cpp

```diff
@@ -278,16 +260,14 @@ static RPCHelpMan waitfornewblock()
         timeout = request.params[0].getInt<int>();
     if (timeout < 0) throw JSONRPCError(RPC_MISC_ERROR, "Negative timeout");
 
-    CUpdatedBlock block;
-    {
-        WAIT_LOCK(cs_blockchange, lock);
-        block = latestblock;
-        if(timeout)
-            cond_blockchange.wait_for(lock, std::chrono::milliseconds(timeout), [&block]() EXCLUSIVE_LOCKS_REQUIRED(cs_blockchange) {return latestblock.height != block.height || latestblock.hash != block.hash || !IsRPCRunning(); });
-        else
-            cond_blockchange.wait(lock, [&block]() EXCLUSIVE_LOCKS_REQUIRED(cs_blockchange) {return latestblock.height != block.height || latestblock.hash != block.hash || !IsRPCRunning(); });
-        block = latestblock;
+    NodeContext& node = EnsureAnyNodeContext(request.context);
+    Mining& miner = EnsureMining(node);
+
+    auto block{CHECK_NONFATAL(miner.getTip()).value()};
+    if (IsRPCRunning()) {
+        block = timeout ? miner.waitTipChanged(block.hash, std::chrono::milliseconds(timeout)) : miner.waitTipChanged(block.hash);
```

Comment:
In commit b4d37fa70506129ba4a4d218bc49ce8ca9d8380c

I think there is a significant change in behaviour here between `waitTipChanged`, or rather `g_best_block`, and the rpc notification through `blockTip`. AFAICT the `blockTip` notification reports on completed operations in `ActivateBestChain` and `InvalidatedBlock`, while `g_best_block` is updated through the lower-level `ConnectTip` and `DisconnectTip`, which may respectively be called iteratively by `ActivateBestChain` and `InvalidateBlock`. So for operations involving multiple blocks this might yield different results here. I'm not sure if these RPCs should report on these more ephemeral state changes rather than the results of the respective operations.

I have not actually tested this, so I might be mistaken here, in which case I think it would still be good to have a note in the commit message comparing the two methods.

Reply:
`g_best_block_cv.notify_all();` is called by `Chainstate::UpdateTip()`, which indeed is called by `ConnectTip` and `DisconnectTip()`. Still wrapping my head around what the previous approach was. I doubt these RPC's were used outside of tests, but it would make sense to document the different.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/30409#discussion_r1727105521

This is a good catch, but I"m not sure there is a observable change of behavior in `ActivateBestChain` because cs_main is held the whole time `blockTip` / `uiInterface.NotifyBlockTip` / `RPCNotifyBlockChange` is called and while the old `latestblock` global (now replaced by `g_best_block`) was updated:

https://github.com/bitcoin/bitcoin/blob/5ce2285b8739e12eebd1d53358038ec2277c662b/src/validation.cpp#L3552

So I don't think `waitTipChanged` will actually return during any intermediate tip changes, because it can't return until it locks `cs_main`, so it should just report the latest block as before.

However, `InvalidBlock` seems like a different story because it is releasing cs_main between disconnecting blocks:

https://github.com/bitcoin/bitcoin/blob/5ce2285b8739e12eebd1d53358038ec2277c662b/src/validation.cpp#L3701

and it would previously only notify about a tip change and update `latestblock` at the end:

https://github.com/bitcoin/bitcoin/blob/5ce2285b8739e12eebd1d53358038ec2277c662b/src/validation.cpp#L3783

But now it will update `g_best_block` and release `cs_main` while it is disconnecting blocks. This seems like something worth documenting, but probably unlikely to cause any problems.

Reply:
Couple of months ago I was thinking how to get rid of `g_best_block` and its friends entirely in the context of the kernel library. We already have two interfaces for notifying if something changed during validation, having a third one, that is also globally mutable and potentially dangerous if multiple chainstate managers are used is not ideal. I did not make progress then, but looking at it again now, I think this might be the opportunity to get rid of it.

How about duplicating the best block logic into the kernel notifications like I've done here https://github.com/TheCharlatan/bitcoin/commit/3c665065f132fd80ea3b1733785b98e45c8b371e (our locking conventions still puzzle me a bit, maybe there is a better way to do it)? This would avoid any change in behavior altogether and mostly make this a refactor. In a follow-up the remaining users of `g_best_block` could then be migrated and the global deleted.

I think the current user of `g_best_block` (getblocktemplate() in mining.cpp), could be migrated to this interface as well. I don't really see the downside of not being notified on these interim states, since it is undesirable to build work on top of them.

Reply:
@TheCharlatan do you want to make a separate PR for the kernel notification changes, or should I just include them here?

Reply:
I think it would be good to integrate them here, and I'll try and follow up with removing the `m_best_header` global from validation.

Reply:
If we are going to add a new `m_tip_block` variable to replace `g_best_block` seems like it would be good to make replacement complete and eliminate `g_best_block`. I think following could work (compiles but untested):

```diff
--- a/src/rpc/mining.cpp
+++ b/src/rpc/mining.cpp
@@ -739,7 +739,6 @@ static RPCHelpMan getblocktemplate()
     {
         // Wait to respond until either the best block changes, OR a minute has passed and there are more transactions
         uint256 hashWatchedChain;
-        std::chrono::steady_clock::time_point checktxtime;
         unsigned int nTransactionsUpdatedLastLP;
 
         if (lpval.isStr())
@@ -760,19 +759,14 @@ static RPCHelpMan getblocktemplate()
         // Release lock while waiting
         LEAVE_CRITICAL_SECTION(cs_main);
         {
-            checktxtime = std::chrono::steady_clock::now() + std::chrono::minutes(1);
-
-            WAIT_LOCK(g_best_block_mutex, lock);
-            while (g_best_block == hashWatchedChain && IsRPCRunning())
-            {
-                if (g_best_block_cv.wait_until(lock, checktxtime) == std::cv_status::timeout)
-                {
-                    // Timeout: Check transactions for update
-                    // without holding the mempool lock to avoid deadlocks
-                    if (miner.getTransactionsUpdated() != nTransactionsUpdatedLastLP)
-                        break;
-                    checktxtime += std::chrono::seconds(10);
-                }
+            MillisecondsDouble checktxtime{std::chrono::minutes(1)};
+            while (tip == hashWatchedChain && IsRPCRunning()) {
+                tip = miner.waitTipChanged(hashWatchedChain, checktxtime).hash;
+                // Timeout: Check transactions for update
+                // without holding the mempool lock to avoid deadlocks
+                if (miner.getTransactionsUpdated() != nTransactionsUpdatedLastLP)
+                    break;
+                checktxtime = std::chrono::seconds(10);
             }
         }
         ENTER_CRITICAL_SECTION(cs_main);
--- a/src/rpc/server.cpp
+++ b/src/rpc/server.cpp
@@ -303,7 +303,6 @@ void StopRPC(const std::any& context)
         LogPrint(BCLog::RPC, "Stopping RPC\n");
         WITH_LOCK(g_deadline_timers_mutex, deadlineTimers.clear());
         DeleteAuthCookie();
-        g_best_block_cv.notify_all();
         Assert(EnsureAnyNodeContext(context).notifications)->m_tip_block_cv.notify_all();
         LogPrint(BCLog::RPC, "RPC stopped.\n");
     });
--- a/src/test/validation_chainstate_tests.cpp
+++ b/src/test/validation_chainstate_tests.cpp
@@ -4,6 +4,7 @@
 //
 #include <chainparams.h>
 #include <consensus/validation.h>
+#include <node/kernel_notifications.h>
 #include <random.h>
 #include <rpc/blockchain.h>
 #include <sync.h>
@@ -69,14 +70,14 @@ BOOST_AUTO_TEST_CASE(validation_chainstate_resize_caches)
 BOOST_FIXTURE_TEST_CASE(chainstate_update_tip, TestChain100Setup)
 {
     ChainstateManager& chainman = *Assert(m_node.chainman);
-    uint256 curr_tip = ::g_best_block;
+    uint256 curr_tip = m_node.notifications->m_tip_block;
 
     // Mine 10 more blocks, putting at us height 110 where a valid assumeutxo value can
     // be found.
     mineBlocks(10);
 
     // After adding some blocks to the tip, best block should have changed.
-    BOOST_CHECK(::g_best_block != curr_tip);
+    BOOST_CHECK(m_node.notifications->m_tip_block != curr_tip);
 
     // Grab block 1 from disk; we'll add it to the background chain later.
     std::shared_ptr<CBlock> pblockone = std::make_shared<CBlock>();
@@ -91,15 +92,15 @@ BOOST_FIXTURE_TEST_CASE(chainstate_update_tip, TestChain100Setup)
     // Ensure our active chain is the snapshot chainstate.
     BOOST_CHECK(WITH_LOCK(::cs_main, return chainman.IsSnapshotActive()));
 
-    curr_tip = ::g_best_block;
+    curr_tip = m_node.notifications->m_tip_block;
 
     // Mine a new block on top of the activated snapshot chainstate.
     mineBlocks(1);  // Defined in TestChain100Setup.
 
     // After adding some blocks to the snapshot tip, best block should have changed.
-    BOOST_CHECK(::g_best_block != curr_tip);
+    BOOST_CHECK(m_node.notifications->m_tip_block != curr_tip);
 
-    curr_tip = ::g_best_block;
+    curr_tip = m_node.notifications->m_tip_block;
 
     BOOST_CHECK_EQUAL(chainman.GetAll().size(), 2);
 
@@ -138,7 +139,7 @@ BOOST_FIXTURE_TEST_CASE(chainstate_update_tip, TestChain100Setup)
     // g_best_block should be unchanged after adding a block to the background
     // validation chain.
     BOOST_CHECK(block_added);
-    BOOST_CHECK_EQUAL(curr_tip, ::g_best_block);
+    BOOST_CHECK_EQUAL(curr_tip, m_node.notifications->m_tip_block);
 }
 
 BOOST_AUTO_TEST_SUITE_END()
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -107,10 +107,6 @@ const std::vector<std::string> CHECKLEVEL_DOC {
  * */
 static constexpr int PRUNE_LOCK_BUFFER{10};
 
-GlobalMutex g_best_block_mutex;
-std::condition_variable g_best_block_cv;
-uint256 g_best_block;
-
 const CBlockIndex* Chainstate::FindForkInGlobalIndex(const CBlockLocator& locator) const
 {
     AssertLockHeld(cs_main);
@@ -2987,12 +2983,6 @@ void Chainstate::UpdateTip(const CBlockIndex* pindexNew)
         m_mempool->AddTransactionsUpdated(1);
     }
 
-    {
-        LOCK(g_best_block_mutex);
-        g_best_block = pindexNew->GetBlockHash();
-        g_best_block_cv.notify_all();
-    }
-
     std::vector<bilingual_str> warning_messages;
     if (!m_chainman.IsInitialBlockDownload()) {
         const CBlockIndex* pindex = pindexNew;
--- a/src/validation.h
+++ b/src/validation.h
@@ -85,11 +85,6 @@ enum class SynchronizationState {
     POST_INIT
 };
 
-extern GlobalMutex g_best_block_mutex;
-extern std::condition_variable g_best_block_cv;
-/** Used to notify getblocktemplate RPC of new tips. */
-extern uint256 g_best_block;
-
 /** Documentation for argument 'checklevel'. */
 extern const std::vector<std::string> CHECKLEVEL_DOC;
 
```

Reply:
Thanks ryanofsky, that is exactly what I had in mind for a potential followup - happy to have this in one go. I still need to think through the trade-offs for the block template rpc, but I think this might actually be better.